File: choitaejin/week4/boj_4779.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

function cantor(arr, start, length, depth) {
  if (depth === 0) {
    return;
  }
  
  const segment = Math.floor(length / 3);
  
  for (let i = start + segment; i < start + 2 * segment; i++) {
    arr[i] = ' ';
  }
  
  cantor(arr, start, segment, depth - 1);
  
  cantor(arr, start + 2 * segment, segment, depth - 1);
}

const inputs = [];

rl.on('line', (line) => {
  if (line === '') {
    rl.close();
    return;
  }
  
  inputs.push(parseInt(line));
}).on('close', () => {
  inputs.forEach(n => {
    const length = Math.pow(3, n);
    const arr = new Array(length).fill('-');
    
    cantor(arr, 0, length, n);
    
    console.log(arr.join(''));
  });
  
  process.exit();
});

File: ojuno/week4/boj_4779.cpp
#include <iostream>
#include <cmath>
#include <string>
using namespace std;

void cantor(string &line, int start, int length) {
    if (length == 1) return;

    int third = length / 3;

    for (int i = start + third; i < start + 2 * third; i++) {
        line[i] = ' ';
    }

    cantor(line, start, third);
    cantor(line, start + 2 * third, third);
}

int main() {
    int n;
    while (cin >> n) {
        int size = pow(3, n);
        string line(size, '-');

        cantor(line, 0, size);
        cout << line << '\n';
    }
    return 0;
}
